src/main/java/ua_parser/CachingParser.java

```diff
@@ -19,105 +22,94 @@
 
-    @SuppressWarnings("unchecked")
     @Override
     public Client parse(String agentString) {
-        if (agentString == null) {
+        if (Strings.isNullOrEmpty(agentString))
             return null;
-        }
-        if (cacheClient == null) {
-            cacheClient = new LRUMap(CACHE_SIZE);
-        }
-        Client client = cacheClient.get(agentString);
-        if (client != null) {
-            return client;
-        }
-        client = super.parse(agentString);
-        cacheClient.put(agentString, client);
-        return client;
+
+        return cacheClient.getUnchecked(agentString);
     }
 
     // ------------------------------------------
 
-    @SuppressWarnings("unchecked")
     @Override
     public UserAgent parseUserAgent(String agentString) {
-        if (agentString == null) {
+        if (Strings.isNullOrEmpty(agentString))
             return null;
-        }
-        if (cacheUserAgent == null) {
-            cacheUserAgent = new LRUMap(CACHE_SIZE);
-        }
-        UserAgent userAgent = cacheUserAgent.get(agentString);
-        if (userAgent != null) {
-            return userAgent;
-        }
-        userAgent = super.parseUserAgent(agentString);
-        cacheUserAgent.put(agentString, userAgent);
-        return userAgent;
+
+        return cacheUserAgent.getUnchecked(agentString);
     }
 
     // ------------------------------------------
 
-    @SuppressWarnings("unchecked")
     @Override
     public Device parseDevice(String agentString) {
-        if (agentString == null) {
+        if (Strings.isNullOrEmpty(agentString))
             return null;
-        }
-        if (cacheDevice == null) {
-            cacheDevice = new LRUMap(CACHE_SIZE);
-        }
-        Device device = cacheDevice.get(agentString);
-        if (device != null) {
-            return device;
-        }
-        device = super.parseDevice(agentString);
-        cacheDevice.put(agentString, device);
-        return device;
+
+        return cacheDevice.getUnchecked(agentString);
     }
 
     // ------------------------------------------
 
-    @SuppressWarnings("unchecked")
     @Override
     public OS parseOS(String agentString) {
-        if (agentString == null) {
+        if (Strings.isNullOrEmpty(agentString))
             return null;
-        }
-
-        if (cacheOS == null) {
-            cacheOS = new LRUMap(CACHE_SIZE);
-        }
-        OS os = cacheOS.get(agentString);
-        if (os != null) {
-            return os;
-        }
-        os = super.parseOS(agentString);
-        cacheOS.put(agentString, os);
-        return os;
+
+        return cacheOS.getUnchecked(agentString);
     }
 
     // ------------------------------------------
+
+    private <T> LoadingCache<String, T> createCache(Function<String, T> loader) {
+        return CacheBuilder.newBuilder()
+                .maximumSize(cacheSize)
+                .expireAfterAccess(expireAfterAccess, TimeUnit.MILLISECONDS)
```

Comment:
count stats (we'd like to monitor this I think) 
in order to get the stats themselves - you'll need to expose the caches with `get` methods